### DISCUSSION TITLE
Deluge save file parsing in the browser

### DISCUSSION THREAD

**Discussion by MrHaila** [SynthstromAudible/DelugeFirmware#94] [Q&A] [answered]
Hey all,

I've been building Deluger, a browser-based Deluge file browser, as a rainy-day hobby. My sincere thanks to @jamiefaye and everyone else who have kindly made their past code available for reference.

On to my issue: An ongoing frustration has been the arcane nature of Deluge's file format. The latest version is much more reasonable to parse than the old ones, but my main issue is how the current file format has _implicit defaults depending on the domain of use_. Outside of proposing any rewrites to the file format, I've been staring at the code in, say, `sound.cpp` to figure out how to replicate this stuff in the browser. My end goal is to reliably parse a file into memory to be able to build a patch inspector UI and possibly even be able to write patches back into files.

I'm writing this in the open to bait for better ideas on how to approach the effort. One (very real) concern is maintenance over time, so I'm not very enthused by the idea of manually converting the +4000 lines of C++ for sounds alone, as that would probably have bugs and bit-rot fairly quickly.

That got me thinking: would it be feasible to compile some of the native C++ code into WASM for the sole purpose of running the exact same serialization code in the browser? This would be a very maintainable way to share the serialization logic between projects. Maybe someone else has a better handle on if the code is written in a way that allows for this.

Failing that, would anyone else be interested in collaborating on a more portable & reusable (typescript?) implementation of the file parsing logic? We could properly document it as a side product 🙂

**Comment by sapphire-arches** [comment]
I think there's a lot of value in fixing up at least the core sequencing and audio code to run on arbitrary platforms. @robmccoll was working on an initial x86 compile target [here](https://github.com/robmccoll/DelugeFirmware/tree/x86_64/compile_tgt) and much of that work can probably be reused toward a WASM target. Putting some layer of abstraction between the sequencing and audio core and the hardware is probably desirable, both for your proposed usecase (external apps that want to interact with the Deluge's file formats) and for testing purposes.

**Reply by robmccoll** [reply]
You could definitely compile a lot of the code base to WASM (like the file parsing logic). Immediate issues I can think of would be emulating things like timers and interrupts. The way I was going to do that on x86 was to use threads and signals. I know there's some version of threads via workers in WASM, but I don't know how you could interrupt control flow. Might be able to get close on the JS side with its timers and just hoping that there aren't any loops that would starve the cooperative scheduling.

**Reply by robmccoll** [reply]
Unfortunately work has really picked up, so I haven't made much progress over the last week or so. I do have a fun method of hooking register reads and writes locally without changing the code that accesses them by abusing operator overloading that I find amusing.

**Comment by PaulFreund** [comment, marked as answer]
I have worked a bit with the XML implementation regarding both parsing and generating (see storageManager) and from what I have seen it is pretty tightly coupled in with the application (sub-parsing handled by child object classes, embedded specific memory management etc.). Also the parsing and generating only allows a very narrow subset of the XML specification so I can understand it is not so easy to generate.

In order to improve the situation it might make sense to refactor the loading and storing to use an embedded compatible XML engine (also to keep compatibility) and also do a little bit of separation between serialization and application code. That code could be written as a module that can also be used by other applications.

**Reply by MrHaila** [reply]
I looked into the code as well and reached the same conclusion. The codebase is too deeply object-oriented for the serialization logic to be "just used" without emulating the whole runtime. A refactor of serialization into its own utility library would indeed sound like a very sensible effort for both maintainability, testability and portability. Unfortunately, I'm not a good enough programmer to write idiomatic C++ libraries, so that effort would have to be completed by someone else 🙁

I'll probably continue thinking about how to maybe export/scrape/type all the domain-specific default values for the different types of objects. That's still a minimum viable strategy to initialize a default object and then read in the XML values to override things. Maybe good enough for UI development.

**Comment by jamiefaye** [comment]
My answer to this problem (in Downrush) was to parse the XML to JSON and
work with that. When I saved back, I reversed the process. I ran into
issues where the XML format was non-standard and so I had to create a
partial ordering table tell the JSON to XML converter what order to write
the elements on it. I based the process on a scan of all the SONG, KIT, and
SYNTH files I could find. One pleasant upshot of this was that it was
super-easy to adapt when Rohan went from nested XML elements to XML
attributes a while back.

While I did not always get a sensible object hierarchy out of this, I did
get something that was relatively robust with regards to most schema
changes.

I have used a similar "scan everything you can find" approach to converting
from Ableton & MPC 2 file formats.

At some point we will define some SysEx messages that let you read and
change Deluge objects. An advantage of this is that you can use Web Midi to
control your Deluge from a web application served from anywhere that has a
SSL certificate. Downrush did not have SSL so it could not use the browser
microphone or SysEx for that matter.

While we are still working out SysEx, I suspect it will do both file-level
and object-level state transfers.

– Jamie

**Reply by MrHaila** [reply]
Thanks for the explanation. I actually did look through your code for this and the approach is clever.

I ended up taking a different approach by coming up with strict typings for the parsed files (separate parsing modes for the firmware versions) and started hitting walls because of the implicit defaults thing. I'd love to be able to show the full configuration of a patch and currently that implies knowing what the defaults are first and then overriding those defaults with whatever is explicitly set in the XML. I currently approximate that by having exported one init synth patch and always loading that before loading the actual XML. Super unscientific, but it works. Now that the sources are out, I guess I might at least look at how the defaults differ in the various domains of use 🤷‍♂️

The SysEx approach is a really good and interesting point. I'd be super happy to write some code against that effort as it progresses! SSL is no problem as, for example, [Delugr](https://delugr.haila.fi/) also needs it for the file system access API and this stuff is easy to host for basically free by anyone. Object-level transfers would be especially interesting, for obvious reasons.

Thanks for the answer and best of luck with taming the USB interface 🙂
